[ new ] Fewer postulates needed for Primitive types

gallais opened the pull request:
Looking at our attempt to make (most of) the standard library compile with the `--safe` flag
(cf. agda/agda-stdlib/pull/414), it would be nice to be able to drop some of the `postulate`s
in `Agda.Builtin.*`.

This is what I came up with. It does allow me to typecheck all of our target modules
in safe mode.

But seeing how some of these builtins trigger [special cases](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Rules/Builtin.hs#L791) inserting extra (invisible) pragmas,
I can't say for sure I haven't broken anything. Opinions?

gallais committed:
[ new ] Fewer postulates needed for Primitive types

gallais committed:
[ fix ] some test cases

gallais commented:
Based on the tests, it looks like we *do* miss some of these silent pragmas.
Here is the error raised by the compilation test case `Succeed/ForeignPragma`:

```
test/Common/IO.agda:36,1-49
The type String → IO ⊤ cannot be translated to a corresponding
Haskell type, because it contains String which does not have a
COMPILE pragma.
Possible fix: add a pragma
  {-# COMPILE GHC String = type HsT #-}
for a suitable Haskell type HsT.
```

A solution would be to add a field for a `QName -> TCM ()` to `BuiltinInfo` and
run that when the `BUILTIN` is declared.

UlfNorell reviewed (changes requested):
The special cases should simply be moved to `bindBuiltinNoDef`, no? Otherwise this looks sensible.

gallais committed:
[ fix ] add compilation Pragmas for base types

gallais committed:
[ warning ] when a BUILTIN rebinds an existing name

This should make our change nicely backwards compatible: we don't
need to postulate String, Char, Float, etc. anymore but if we are
then Agda simply warns us.

gallais commented:
Current problem: BUILTIN-bound names are assigned `noFixity` which means that `♯_`
gets the wrong level. I should be able to check whether there is a pre-existing fixity
declaration and use that instead.

I'll look into it after lunch.

gallais committed:
Merge branch 'master' into builtin

UlfNorell commented on src/full/Agda/TypeChecking/Rules/Builtin.hs:
```diff
@@ -857,28 +853,37 @@ bindUntypedBuiltin b = \case
-bindBuiltinNoDef' b q = inTopContext $ do
-  case builtinDesc <$> findBuiltinInfo b of
+bindBuiltinNoDef' b q = do
+   case builtinDesc <$> findBuiltinInfo b of
```
Accidental space!! 🙀

gallais committed:
[ fixity ] re-introduce fixity declaration for flat

* remove the warning: BUILTIN pragmas for a builtinsNoDef name
  are now recognized as proper declarations.

* attach fixity information to BUILTIN-declared identifiers: In the
  parser we use __IMPOSSIBLE__ rather than noFixity' as the default
  value so that we don't silently succeed if we forget to grab the
  right fixity in the nicifier

gallais commented on src/full/Agda/Syntax/Translation/AbstractToConcrete.hs:
```diff
@@ -952,10 +952,10 @@ data RangeAndPragma = RangeAndPragma Range A.Pragma
+    A.BuiltinNoDefPragma b x   -> C.BuiltinPragma r b <$> toConcrete x <*> pure (nameFixity $ qnameName x)
+    A.RewritePragma x          -> C.RewritePragma r . singleton <$> toConcrete x
+    A.CompiledTypePragma x hs  -> do
       x <- toConcrete x
```
grrr another extra space

gallais committed:
[ cosmetic ] removing useless space

[ci skip]

gallais committed:
[ fix ] updating golden values for test cases

gallais committed:
[ haddock ] disallows bullet points in the comments

gallais commented:
The [previous build](https://travis-ci.org/agda/agda/builds/416906115) succeeded for everything except haddock so I've cancelled all the jobs
[except for the haddock one](https://travis-ci.org/agda/agda/builds/416961319) on this one. If it is successful then we're good to go!

andreasabel commented:
Well, this is in part to be blamed on me, since I started builtins without postulates for sized types.  There, the rationale was that things like `SizeUniv` cannot be typed.  At that time, someone remarked/complained that pragmas cannot be ignored any longer in reading code since they add symbols to the scope.
With the present PR, we seem to have abandoned the "pragmas are just technical detail to be ignored on first reading" altogether.  It would be consequent to add a proper syntax for builtins now that does not look like a comment, similar to `primitive`.
```agda
builtin
  STRING String
  EQUALITY _==_
```
